File: programming/4.1/main.cpp
#include "../include/VariadicTable.h"
#include <iostream>
#include <cmath>
#include <string>


using namespace std;

int randint(int min, int max) {
  return rand() % (max + 1 - min) + min;
}

float calculateIncome(int k, int i) {
  return k * randint(-i * randint(0, 50), 100);
}

int main() {
  srand(time(NULL));
  VariadicTable<int, float, string> vt({"year", "income", "tag"}, 2022 - 1991);
  int maxIncomeYear = 0, maxIncomeValue = 0;
  int minIncomeYear = 0, minIncomeValue = 0;
  int income;
  int k = 1;
  string tag;


  for (int i = 1991; i < 2022; i++) {
    k = i % 10;

    income = calculateIncome(k, 16);
    if (income > 0) {
      tag = "income";
    }
    else {
      tag = "loss";
    }

    if (income < 0) {
      if (income < minIncomeValue) {
        minIncomeYear = i;
        minIncomeValue = income;
      }
    }

    if (income > 0) {
      if (income > maxIncomeValue) {
        cout << i << endl;
        maxIncomeYear = i;
        maxIncomeValue = income;
      }
    }
    vt.addRow(i, income, tag);
  }

  vt.print(cout);

  return 0;
}

File: programming/4.1/main.py
from random import randint
from prettytable import PrettyTable


# define table
table = PrettyTable() 
table.field_names = ['year', 'income', 'tag']

# define variables
table_data_item_default = {
  "year": 0,
  "income": 0,
  "tag": "",
}
table_data = []
income = 0
tag = ''

def calculateIncome(k, i) -> float:
  return k * randint(-i * randint(0, 50), 100)

minIncome = 0
maxIncome = 0
k = 0
tag = ''

for i in range(1991, 2022 + 1):
  k = i % 10
  income = calculateIncome(k, i)

  if not minIncome:
    minIncome = income

  if not maxIncome:
    maxIncome = income

  # min
  if income < 0:
    if income < minIncome:
      minIncome = i

  # max
  if income > 0:
    if income > maxIncome:
      maxIncome = i

  # define tag
  if income > 0:
    tag = 'income'
  else:
    tag = 'loss'


  table_data.append({
    "year": i,
    "income": income,
  })

for i in table_data:
  if minIncome == i['year']:
    tag = 'min income'

  elif maxIncome == i['year']:
    tag = 'max income'

  table.add_row([i['year'], i['income'], tag])


print(table)
